src/ORM/Marshaller.php

```diff
@@ -294,12 +294,26 @@ protected function _belongsToMany(Association $assoc, array $data, $options = []
                     $query->orWhere($keys);
                 }
             } else {
-                $records[] = $this->one($row, $options);
+                $records[$i] = $this->one($row, $options);
             }
         }
 
         if (isset($query)) {
-            $records = array_merge($records, $query->toArray());
+            $queryData = $query->toArray();
+            foreach ($queryData as $queryItem) {
+                foreach ($data as $i => $row) {
+                    $matchCount = 0;
+                    foreach ($primaryKey as $key => $value) {
```

Comment:
We might be able to reduce the loop depth by using a Collection to index both $data and the $queryData results by primary key before stitching them together. I can give that a shot.